kv/memberlist/memberlist_client.go

```diff
@@ -568,41 +568,88 @@ func (m *KV) joinMembersOnStartup(ctx context.Context) bool {
 		return true
 	}
 
+	logger := log.With(m.logger, "phase", "startup")
+	level.Info(logger).Log("msg", "joining memberlist cluster", "join_members", strings.Join(m.cfg.JoinMembers, ","))
 	startTime := time.Now()
+	reached, err := m.joinMembersInBatches(ctx, m.cfg.MaxJoinRetries, logger)
+	if err != nil {
+		level.Error(logger).Log("msg", "joining memberlist cluster failed", "err", err, "elapsed_time", time.Since(startTime))
+		return false
+	}
+	level.Info(logger).Log("msg", "joining memberlist cluster succeeded", "reached_nodes", reached, "elapsed_time", time.Since(startTime))
+	return true
+}
 
-	level.Info(m.logger).Log("msg", "joining memberlist cluster", "join_members", strings.Join(m.cfg.JoinMembers, ","))
+// joinMembersInBatches joins m.cfg.JoinMembers 100 at a time. After each batch of 100 it rediscoveres the members.
+// This helps when the list of members is big and by the time we reach the end the originally resolved addresses may be obsolete.
+// joinMembersInBatches returns an error iff it couldn't successfully join any node.
+func (m *KV) joinMembersInBatches(ctx context.Context, numAttempts int, logger log.Logger) (int, error) {
+	const joinNodesBatch = 100
+	var (
+		cfg = backoff.Config{
+			MinBackoff: m.cfg.MinJoinBackoff,
+			MaxBackoff: m.cfg.MaxJoinBackoff,
+			MaxRetries: numAttempts,
+		}
+		boff               = backoff.New(ctx, cfg)
+		lastErr            = error(nil)
+		successfullyJoined = 0
+	)
+
+	for ; boff.Ongoing(); boff.Wait() {
```

Comment:
[nit] To keep the code easier to read, I would move the whole content of the `for` loop to the function `joinMembersInBatches()` and rename this function to `joinMembersInBatchesWithRetries()`. In `joinMembersInBatchers()` instead of checking `boff.Ongoing()` you could just check the `ctx.Err()`.